File: python ds/DoublyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def insert_at_beginning(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        new_node.next = self.head
        self.head.prev = new_node
        self.head = new_node

    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node
        new_node.prev = temp

    def delete_node(self, key):
        temp = self.head
        if temp is None:
            return
        
        while temp and temp.data != key:
            temp = temp.next
        
        if temp is None:
            return
        
        if temp.prev:
            temp.prev.next = temp.next
        else:
            self.head = temp.next
        
        if temp.next:
            temp.next.prev = temp.prev

    def display_forward(self):
        temp = self.head
        while temp:
            print(temp.data, end=' <-> ')
            temp = temp.next
        print('None')

    def display_backward(self):
        temp = self.head
        if temp is None:
            print("List is empty")
            return
        while temp.next:
            temp = temp.next
        while temp:
            print(temp.data, end=' <-> ')
            temp = temp.prev
        print('None')

    def search(self, key):
        temp = self.head
        while temp:
            if temp.data == key:
                return True
            temp = temp.next
        return False

# Example Usage
dll = DoublyLinkedList()
dll.insert_at_end(10)
dll.insert_at_end(20)
dll.insert_at_beginning(5)
dll.insert_at_end(30)

dll.display_forward()  
dll.display_backward()  

dll.delete_node(20)
dll.display_forward()  

File: C ds/doubly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node{
    int data;
    struct Node* prev;
    struct Node* next;
} Node;

typedef struct {
    Node* head;
} DLL;

Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode -> data = data;
    newNode -> prev = NULL;
    newNode -> next = NULL;
    return newNode;
}

void initialize(DLL* dll) {
    dll -> head = NULL;
}

void insertAtBeginning(DLL* dll, int data) {
    Node* newNode = createNode(data);
    if(dll -> head == NULL) {
        dll -> head = newNode;
        return;
    }
    newNode -> next = dll -> head;
    dll -> head -> prev = newNode;
    dll -> head = newNode;
}

void insertAtEnd(DLL* dll, int data) {
    Node* newNode = createNode(data);
    if(dll -> head == NULL) {
        dll -> head = newNode;
        return;
    }
    Node* temp = dll -> head;
    while(temp -> next != NULL) {
        temp = temp -> next;
    }
    temp -> next = newNode;
    newNode -> prev = temp;
}

void deleteAt(DLL* dll, int key) {
    Node* temp = dll -> head;

    if (temp == NULL){
        printf("List empty!! \n");
        return;
    }

    while(temp != NULL && temp -> data != key) {
        temp = temp -> next;
    }

    if (temp == NULL) {
        printf("Item not found. \n");
        return;
    }

    if (temp == dll->head) {
        dll->head = temp->next;
        if (dll->head) dll->head->prev = NULL;
    } else {
        temp->prev->next = temp->next;
        if (temp->next) temp->next->prev = temp->prev;
    }

    free(temp);
}

void displayForward(DLL* dll) {
    if (dll->head == NULL) {
        printf("List is empty\n");
        return;
    }

    Node* temp = dll -> head;
    printf("Forward: ");
    while(temp != NULL) {
        printf("%d <-> ", temp -> data);
        temp = temp -> next;
    }
    printf("NULL \n");
}

void displayBackward(DLL* dll) {
    if (dll->head == NULL) {
        printf("List is empty\n");
        return;
    }
    Node* temp = dll->head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    printf("Backward: ");
    while (temp != NULL) {
        printf("%d <-> ", temp->data);
        temp = temp->prev;
    }
    printf("NULL\n");
}

int search(DLL* dll, int key) {
    Node* temp = dll -> head;
    while(temp != NULL) {
        if(temp -> next == key) {
            return 1;
        }
    }
    return 0;
}

int main() {

    DLL dll;
    initialize(&dll);

    insertAtEnd(&dll, 10);
    insertAtEnd(&dll, 20);
    insertAtBeginning(&dll, 5);
    insertAtEnd(&dll, 30);

    displayForward(&dll);
    displayBackward(&dll);

    printf("Deleting 20...\n");
    deleteAt(&dll, 20);
    displayForward(&dll);

    printf("Searching for 10: %s\n", search(&dll, 10) ? "Found" : "Not Found");
    printf("Searching for 50: %s\n", search(&dll, 50) ? "Found" : "Not Found");

    return 0;

}
